sdk/python/kfp/compiler/compiler_test.py

```diff
@@ -4480,5 +4543,819 @@ def flip_coin_pipeline():
                     print_and_return(text='Got tails!')
 
 
+class TestDslOneOf(unittest.TestCase):
+    # The space of possible tests is very large, so we test a representative set of cases covering the following styles of usage:
+    # - upstream conditions: if/else v if/elif/else
+    # - data consumed: parameters v artifacts
+    # - where dsl.OneOf goes: consumed by task v returned v both
+    # - when outputs have different keys: e.g., .output v .outputs[<key>]
+    # - how the if/elif/else are nested and at what level they are consumed
+
+    # Data type validation (e.g., dsl.OneOf(artifact, param) fails) and similar is covered in pipeline_channel_test.py.
+
+    # To help narrow the tests further (we already test lots of aspects in the following cases), we choose focus on dsl.OneOf behavior, not the conditional logic if If/Elif/Else. This is more verbose, but more maintainable and the behavior under test is clearer.
+
+    def test_if_else_returned(self):
+        # if/else
+        # returned
+        # parameters
+        # different output keys
+
+        @dsl.pipeline
+        def roll_die_pipeline() -> str:
+            flip_coin_task = flip_coin()
+            with dsl.If(flip_coin_task.output == 'heads'):
+                t1 = print_and_return(text='Got heads!')
+            with dsl.Else():
+                t2 = print_and_return_with_output_key(text='Got tails!')
+            return dsl.OneOf(t1.output, t2.outputs['output_key'])
+
+        # hole punched through if
+        self.assertEqual(
+            roll_die_pipeline.pipeline_spec.components['comp-condition-2']
+            .output_definitions.parameters[
+                'pipelinechannel--print-and-return-Output'].parameter_type,
+            type_utils.STRING,
+        )
+        # hole punched through else
+        self.assertEqual(
+            roll_die_pipeline.pipeline_spec.components['comp-condition-3']
+            .output_definitions.parameters[
+                'pipelinechannel--print-and-return-with-output-key-output_key']
+            .parameter_type,
+            type_utils.STRING,
+        )
+        # condition-branches surfaces
+        self.assertEqual(
+            roll_die_pipeline.pipeline_spec
+            .components['comp-condition-branches-1'].output_definitions
+            .parameters['pipelinechannel--condition-branches-1-oneof-1']
+            .parameter_type,
+            type_utils.STRING,
+        )
+        parameter_selectors = roll_die_pipeline.pipeline_spec.components[
+            'comp-condition-branches-1'].dag.outputs.parameters[
+                'pipelinechannel--condition-branches-1-oneof-1'].value_from_oneof.parameter_selectors
+        self.assertEqual(
+            parameter_selectors[0].output_parameter_key,
+            'pipelinechannel--print-and-return-Output',
+        )
+        self.assertEqual(
+            parameter_selectors[0].producer_subtask,
+            'condition-2',
+        )
```

Comment:
Thanks, I like that suggestion. I applied it in places where it would lead to less code or more readable code. I opted not to for other places where reconstructing the whole object would be unnecessary for the sake of the code under test.